feat: add RequestCommunityInfoFromMailserverAsync

jrainville opened the pull request:
This function doesn't wait for the response from mailservers, so the client has to rely on signals

This was added because the basic `RequestCommunityInfoFromMailserver` could take a lot of time to resolve, and it would either freeze the UI or, in the caser of a different thread, prevent the app from closing since the thread wouldn't close until the function finished.

I also remove `RequestCommunityInfoFromMailserverSync`, because it was literally just calling "normal" `RequestCommunityInfoFromMailserver`. So places that called `RequestCommunityInfoFromMailserverSync` now call `RequestCommunityInfoFromMailserver`.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

jrainville commented:
This new endpoint is used in this desktop PR https://github.com/status-im/status-desktop/pull/4793

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (6)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3cc80588 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2537/1/) | 2022-02-14 17:24:49 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220214-172227-3cc805-pr2537.zip) |
| :heavy_check_mark: | 3cc80588 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2537/1/) | 2022-02-14 17:26:33 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220214-172222-3cc805-pr2537.zip) |
| :heavy_check_mark: | 3cc80588 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2537/1/) | 2022-02-14 17:27:02 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220214-172222-3cc805-pr2537.aar) |
| | | | | | | |
| :heavy_check_mark: | 6c47e208 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2537/2/) | 2022-02-14 18:25:33 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220214-182251-6c47e2-pr2537.zip) |
| :heavy_check_mark: | 6c47e208 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2537/2/) | 2022-02-14 18:25:42 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220214-182251-6c47e2-pr2537.zip) |
| :heavy_check_mark: | 6c47e208 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2537/2/) | 2022-02-14 18:27:42 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220214-182251-6c47e2-pr2537.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 92f9651c | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2537/3/) | 2022-02-14 21:28:42 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220214-212637-92f965-pr2537.zip) |
| :heavy_check_mark: | 92f9651c | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2537/3/) | 2022-02-14 21:29:50 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220214-212632-92f965-pr2537.zip) |
| :heavy_check_mark: | 92f9651c | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2537/3/) | 2022-02-14 21:30:49 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220214-212632-92f965-pr2537.aar) |
| | | | | | | |
| :heavy_check_mark: | 8207f2d9 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2537/4/) | 2022-02-14 21:38:47 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220214-213552-8207f2-pr2537.zip) |
| :heavy_check_mark: | 8207f2d9 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2537/4/) | 2022-02-14 21:39:00 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220214-213552-8207f2-pr2537.zip) |
| :heavy_check_mark: | 8207f2d9 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2537/4/) | 2022-02-14 21:39:36 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220214-213552-8207f2-pr2537.aar) |

cammellos commented on protocol/messenger_communities.go:
```diff
@@ -746,21 +746,21 @@ func (m *Messenger) BanUserFromCommunity(request *requests.BanUserFromCommunity)
+// RequestCommunityInfoFromMailserverAsync installs filter for community and requests its details
+// from mailserver. When response received it will be passed through signals handler
+func (m *Messenger) RequestCommunityInfoFromMailserverAsync(communityID string) error {
+	_, err := m.requestCommunityInfoFromMailserver(communityID, true)
```
Why not do the opposite so we don't have to negate it? so passing here false, above true, and then you don't need to swap `async`, we could call it `waitForResponse` as it's called in the code, what do you think?

jrainville replied:
Fixed

cammellos commented on protocol/messenger_communities.go:
```diff
@@ -792,12 +792,16 @@ func (m *Messenger) requestCommunityInfoFromMailserver(communityID string) (*com
 		return nil, err
 	}
 
+	if async {
```
I think you still want to return `err`, if the call is async it will return early, but say there's an error and the call is not made at all, you probably want to be notified

jrainville replied:
The error check is done already on the lines above (797), unless I misunderstood your point?

jrainville replied:
Updated to merge the `err` and `waitForResponse` checks

cammellos replied:
oh wait, maybe I did miss something, but basically, always return the error

jrainville committed:
feat: add RequestCommunityInfoFromMailserverAsync

This function doesn't wait for the response from mailservers, so the client has to rely on signals

cammellos commented on protocol/messenger_communities.go:
```diff
@@ -792,9 +792,9 @@ func (m *Messenger) requestCommunityInfoFromMailserver(communityID string) (*com
+		waitForResponse)
 
-	if err != nil {
+	if err != nil || !waitForResponse {
```
here you want to always return the error, sorry for the confusion if I asked you to change it before, I must have misread it.

Essentially:


``` 
if err != nil {
  return nil, err
}

if !waitForResponse {
return nil, nil
}

```

cammellos replied:
or I guess it's the same as you wrote, I need to go to bed :) nevermind

jrainville replied:
Alright, it's back to the way it was before, the way you commented

osmaczko commented on protocol/messenger_communities.go:
```diff
@@ -792,12 +792,16 @@ func (m *Messenger) requestCommunityInfoFromMailserver(communityID string) (*com
 	}
 
+	if !waitForResponse {
+		return nil, nil
```
In this case, the filter for this community will not be removed after the community is fetched. Is that desired behavior?
